File: src/Domain.LinnApps/Stock/StorageLocation.cs
namespace Linn.Stores2.Domain.LinnApps.Stock
{
    using System;
    using Linn.Stores2.Domain.LinnApps.Accounts;
    using Linn.Stores2.Domain.LinnApps.Exceptions;

    public class StorageLocation
    {
        public StorageLocation()
        {
            // for ef       
        }

        public StorageLocation(
            int locationId,
            string locationCode,
            string description,
            StorageSite site,
            StorageArea area,
            AccountingCompany company,
            string accessible,
            string storesKittable,
            string mixStates,
            string stockState,
            string typeOfStock,
            StockPool stockPool,
            StorageType storageType)
        {
            this.LocationId = locationId;
            this.LocationCode = locationCode;
            this.Description = description;
            if (site == null)
            {
                throw new StorageLocationException("Location needs a site");
            }

            this.SiteCode = site.SiteCode;
            
            if (area == null)
            {
                throw new StorageLocationException("Location needs an area");
            }

            if (!string.IsNullOrEmpty(site.SitePrefix) || !string.IsNullOrEmpty(area.AreaPrefix))
            {
                var prefixShouldBe = $"{site.SitePrefix}-{area.AreaPrefix}-";
                if (!locationCode.StartsWith(prefixShouldBe))
                {
                    throw new StorageLocationException($"Cannot create Location - Location code should start with {prefixShouldBe}");
                }
            }

            this.StorageAreaCode = area.StorageAreaCode;
            this.StorageArea = area;

            if (company == null)
            {
                throw new StorageLocationException("Location needs an accounting company");
            }

            this.AccountingCompany = company.Name;

            this.CheckYesNoFlag(accessible, "Cannot create Location - accessible should be Y or N");
            this.AccessibleFlag = accessible;

            this.CheckYesNoFlag(storesKittable, "Cannot create Location - stores kittable should be Y, N or blank", true);
            this.StoresKittableFlag = storesKittable;

            this.CheckYesNoFlag(mixStates, "Cannot create Location - mix states should be Y or N");
            this.MixStatesFlag = mixStates;

            this.CheckThreeValueFlag(stockState, "I", "Q", "A", "Cannot create Location - stock state should be I, Q or A");
            this.StockState = stockState;

            this.CheckThreeValueFlag(typeOfStock, "A", "R", "F", "Cannot create Location - type of stock should be R, F or A");
            this.TypeOfStock = typeOfStock;

            this.DefaultStockPool = stockPool?.StockPoolCode;

            this.StorageTypeCode = storageType?.StorageTypeCode;
            this.StorageType = storageType;
        }

        public int LocationId { get; set; }

        public string LocationCode { get; set; }

        public string Description { get; set; }

        public string LocationType { get; set; }

        public string DefaultStockPool { get; set; }

        public DateTime? DateInvalid { get; set; }

        public string StorageTypeCode { get; set; }

        public StorageType StorageType { get; set; }

        public string SiteCode { get; set; }

        public string StorageAreaCode { get; set; }

        public StorageArea StorageArea { get; set; }

        public string AccountingCompany { get; set; }

        public int? SalesAccountId { get; set; }

        public int? OutletNumber { get; set; }

        public string MixStatesFlag { get; set; }

        public string StockState { get; set; }

        public string TypeOfStock { get; set; }

        public string SpecProcFlag { get; set; }

        public string AccessibleFlag { get; set; }

        public string StoresKittableFlag { get; set; }

        public int? StoresKittingPriority { get; set; }

        public int? AuditFrequencyWeeks { get; set; }

        public int? AuditedByEmployeeId { get; set; }

        public Employee AuditedBy { get; set; }

        public DateTime? DateLastAudited { get; set; }

        public string AuditedByDepartmentCode { get; set; }

        public Department AuditedByDepartment { get; set; }

        public bool MixStates() => MixStatesFlag == "Y";

        public bool Accessible() => AccessibleFlag == "Y";

        public bool StoresKittable() => StoresKittableFlag == "Y";

        public void Update(string description, AccountingCompany company, string accessible, string storesKittable,
            string mixStates, string stockState, string typeOfStock, StockPool stockPool, StorageType storageType, DateTime? dateInvalid)
        {
            this.Description = description;

            if (company == null)
            {
                throw new StorageLocationException("Location needs an accounting company");
            }

            this.AccountingCompany = company.Name;

            this.CheckYesNoFlag(accessible, "Cannot update Location - accessible should be Y or N");
            this.AccessibleFlag = accessible;

            this.CheckYesNoFlag(storesKittable, "Cannot update Location - stores kittable should be Y, N or blank", true);
            this.StoresKittableFlag = storesKittable;

            this.CheckYesNoFlag(mixStates, "Cannot update Location - mix states should be Y or N");
            this.MixStatesFlag = mixStates;

            this.CheckThreeValueFlag(stockState, "I", "Q", "A", "Cannot update Location - stock state should be I, Q or A");
            this.StockState = stockState;

            this.CheckThreeValueFlag(typeOfStock, "A", "R", "F", "Cannot update Location - type of stock should be R, F or A");
            this.TypeOfStock = typeOfStock;

            this.DefaultStockPool = stockPool?.StockPoolCode;

            this.StorageTypeCode = storageType?.StorageTypeCode;
            this.StorageType = storageType;

            this.DateInvalid = dateInvalid;
        }

        private void CheckYesNoFlag(string field, string errorMessage, bool allowNull = false)
        {
            if (string.IsNullOrEmpty(field) && allowNull)
            {
                return;
            }
            if (!(field == "Y" || field == "N"))
            {
                throw new StorageLocationException(errorMessage);
            }
        }

        private void CheckThreeValueFlag(string field, string value1, string value2, string value3, string errorMessage)
        {
            if (!(field == value1 || field == value2 || field == value3))
            {
                throw new StorageLocationException(errorMessage);
            }
        }
    }
}

File: src/Service.Host/client/src/components/StorageLocation.js
import React, { useEffect, useState } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { useLocation } from 'react-router';
import queryString from 'query-string';
import moment from 'moment';
import List from '@mui/material/List';
import Grid from '@mui/material/Grid2';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';
import {
    Dropdown,
    ErrorCard,
    InputField,
    Loading,
    SaveBackCancelButtons
} from '@linn-it/linn-form-components-library';
import PropTypes from 'prop-types';
import Page from './Page';
import config from '../config';
import itemTypes from '../itemTypes';
import useGet from '../hooks/useGet';
import useInitialise from '../hooks/useInitialise';
import usePut from '../hooks/usePut';
import usePost from '../hooks/usePost';

function StorageLocation({ creating }) {
    const [hasFetched, setHasFetched] = useState(false);
    const location = useLocation();
    const query = queryString.parse(location.search);

    const {  isLoading: sitesLoading, result: sitesResult } = useInitialise(itemTypes.storageSites.url);
    const {  isLoading: accountingCompaniesLoading, result: accountingCompaniesResult } = useInitialise(itemTypes.accountingCompany.url);
    const {  isLoading: stockPoolsLoading, result: stockPoolsResult } = useInitialise(itemTypes.stockPools.url);
    const {  isLoading: storageTypesLoading, result: storageTypesResult } = useInitialise(itemTypes.storageTypes.url);

    const { id } = useParams();
    const {
        send: getLocation,
        isLoading,
        result: locationGetResult
    } = useGet(itemTypes.storageLocations.url);

    const {
        send: updateLocation,
        isLoading: updateLoading,
        errorMessage: updateError,
        putResult: updateResult
    } = usePut(itemTypes.storageLocations.url, true);

    const {
        send: createLocation,
        isLoading: createLoading,
        errorMessage: createError
    } = usePost(itemTypes.storageLocations.url, true, true);

    if (!creating && !hasFetched) {
        setHasFetched(true);
        getLocation(id);
    }

    const navigate = useNavigate();

    const [formValues, setFormValues] = useState();
    const [changesMade, setChangesMade] = useState(false);

    if (locationGetResult && !formValues) {
        setFormValues(locationGetResult);
    }

    if (creating && !formValues) {
        setFormValues({
            description: null,
            siteCode: query?.siteCode ? query?.siteCode : null,
            storageAreaCode: query?.storageAreaCode ? query?.storageAreaCode : null,
            locationCode: query?.prefix ? query?.prefix : null,
            accountingCompany: 'LINN',
            storageType: null,
            mixStatesFlag: 'Y',
            stockState: 'A',
            typeOfStock: 'A',
            specProcFlag: null,
            storesKittable: 'Y',
            storesKittingPriority: null,
            accessibleFlag: 'Y',
            defaultStockPool: null,
            auditFrequencyWeeks: null
        });
    }

    const selectedSite = () => {
        if (formValues.siteCode && sitesResult) {
            return sitesResult.find(s => s.siteCode === formValues.siteCode);
        }
        return null;
    };

    const locationCodePrefix = newAreaCode => {
        const site = selectedSite();
        if (!site) {
            return '';
        }
        const area = site.storageAreas.find(s => s.storageAreaCode === newAreaCode);
        return area ? `${site.sitePrefix}-${area.areaPrefix}-` : '';
    };

    const lastAuditInfo = () => {
        if (locationGetResult) {
            if (locationGetResult.dateLastAudited) {
                return `Last audited ${moment(locationGetResult.dateLastAudited).format('DD MMM YY')} by ${locationGetResult.auditedBy} dept ${locationGetResult.auditedByDepartmentCode} ${locationGetResult.auditedByDepartmentName}`;
            }

            return 'Not audited';
        }
        return null;
    };

    const handleFieldChange = (propertyName, newValue) => {
        if (creating && propertyName === 'storageAreaCode') {
            setFormValues(current => ({
                ...current,
                [propertyName]: newValue,
                locationCode: locationCodePrefix(newValue)
            }));
        } else {
            setFormValues(current => ({ ...current, [propertyName]: newValue }));
        }
        setChangesMade(true);
    };

    const okToSave = () => {
        if (creating) {
            return formValues.siteCode && formValues.storageAreaCode && formValues.accountingCompany && formValues.description;
        }
        return true;
    }

    const makeInvalid = () => {
        handleFieldChange("dateInvalid", new Date());
    }

    useEffect(() => {
        if (updateResult) {
            setFormValues(updateResult);
        }
    }, [updateResult]);

    return (
        <Page homeUrl={config.appRoot} showAuthUi={false}>
            <Grid container spacing={3}>
                <Grid size={12}>
                    <Typography variant="h4">
                        {creating ? 'Create Storage Location' : 'Storage Location'}
                    </Typography>
                </Grid>
                {updateError && (
                    <Grid size={12}>
                        <List>
                            <ErrorCard errorMessage={updateError} />
                        </List>
                    </Grid>
                )}
                {createError && (
                    <Grid size={12}>
                        <List>
                            <ErrorCard errorMessage={createError} />
                        </List>
                    </Grid>
                )}

                {isLoading || updateLoading || createLoading ? (
                    <Grid size={12}>
                        <List>
                            <Loading />
                        </List>
                    </Grid>
                ) : (
                    formValues && (
                        <>
                            <Grid size={5}>
                                {sitesResult && <Dropdown
                                    value={formValues.siteCode}
                                    fullWidth
                                    label="Site"
                                    propertyName="siteCode"
                                    disabled={!creating}
                                    allowNoValue
                                    items={sitesResult?.map(c => ({
                                        id: c.siteCode,
                                        displayText: c.description
                                    }))}
                                    onChange={handleFieldChange}
                                />}
                            </Grid>
                            <Grid size={7}>
                                <Dropdown
                                    value={formValues.storageAreaCode}
                                    fullWidth
                                    label="Area"
                                    propertyName="storageAreaCode"
                                    disabled={!creating}
                                    allowNoValue
                                    items={selectedSite()?.storageAreas?.map(c => ({
                                        id: c.storageAreaCode,
                                        displayText: c.description
                                    }))}
                                    onChange={handleFieldChange}
                                />
                            </Grid>
                            <Grid size={5}>
                                <InputField
                                    disabled={
                                        !creating ||
                                        !formValues.siteCode ||
                                        !formValues.storageAreaCode
                                    }
                                    value={formValues.locationCode}
                                    fullWidth
                                    label="Code"
                                    propertyName="locationCode"
                                    onChange={handleFieldChange}
                                />
                            </Grid>
                            <Grid size={7}>
                                <InputField
                                    value={formValues.description}
                                    fullWidth
                                    label="Description"
                                    propertyName="description"
                                    onChange={handleFieldChange}
                                />
                            </Grid>
                            <Grid size={5}>
                                {accountingCompaniesResult && (
                                    <Dropdown
                                        value={formValues.accountingCompany}
                                        fullWidth
                                        label="Accounting Company"
                                        propertyName="accountingCompany"
                                        allowNoValue
                                        items={accountingCompaniesResult.map(c => ({
                                            id: c.name,
                                            displayText: c.description
                                        }))}
                                        onChange={handleFieldChange}
                                    />
                                )}
                            </Grid>
                            <Grid size={7}>
                                {storageTypesResult && <Dropdown
                                    value={formValues.storageType}
                                    fullWidth
                                    label="Storage Type"
                                    propertyName="storageType"
                                    allowNoValue
                                    items={storageTypesResult.map(c => ({
                                        id: c.storageTypeCode,
                                        displayText: `${c.storageTypeCode} - ${c.description}` 
                                    }))}
                                    onChange={handleFieldChange}
                                />}
                            </Grid>
                            <Grid size={5}>
                                {stockPoolsResult && <Dropdown
                                    value={formValues.defaultStockPool}
                                    fullWidth
                                    label="Default Stock Pool"
                                    propertyName="defaultStockPool"
                                    allowNoValue
                                    items={stockPoolsResult.map(c => ({
                                        id: c.stockPoolCode,
                                        displayText: c.stockPoolDescription
                                    }))}
                                    onChange={handleFieldChange}
                                />}
                            </Grid>
                            <Grid size={4}>
                                <InputField
                                    value={formValues.salesAccountId}
                                    fullWidth
                                    label="Account Id"
                                    propertyName="salesAccountId"
                                    onChange={handleFieldChange}
                                />
                            </Grid>
                            <Grid size={3}>
                                <InputField
                                    value={formValues.outletNumber}
                                    fullWidth
                                    label="Outlet Num"
                                    propertyName="outletNumber"
                                    onChange={handleFieldChange}
                                />
                            </Grid>
                            <Grid size={2}>
                                <Dropdown
                                    value={formValues.mixStatesFlag}
                                    fullWidth
                                    label="Mix States?"
                                    propertyName="mixStatesFlag"
                                    items={[
                                        { id: 'Y', displayText: 'Yes' },
                                        { id: 'N', displayText: 'No' }
                                    ]}
                                    onChange={handleFieldChange}
                                />
                            </Grid>
                            <Grid size={3}>
                                <Dropdown
                                    value={formValues.stockState}
                                    fullWidth
                                    label="Stock State"
                                    propertyName="stockState"
                                    items={[
                                        { id: 'A', displayText: 'Any Stock' },
                                        { id: 'I', displayText: 'Inspected Stock' },
                                        { id: 'Q', displayText: 'QC/Fail Stock' }
                                    ]}
                                    onChange={handleFieldChange}
                                />
                            </Grid>
                            <Grid size={4}>
                                <Dropdown
                                    value={formValues.stockState}
                                    fullWidth
                                    label="Type of Stock"
                                    propertyName="stockState"
                                    items={[
                                        { id: 'A', displayText: 'Any Stock' },
                                        { id: 'R', displayText: 'Raw Materials' },
                                        { id: 'F', displayText: 'Finished Goods' }
                                    ]}
                                    onChange={handleFieldChange}
                                />
                            </Grid>
                            <Grid size={3}>
                                <InputField
                                    value={formValues.salesAccountId}
                                    fullWidth
                                    label="Spec Proc Flag"
                                    propertyName="specProcFlag"
                                    onChange={handleFieldChange}
                                />
                            </Grid>
                            <Grid size={2}>
                                <Dropdown
                                    value={formValues.mixStatesFlag}
                                    fullWidth
                                    label="Stores Kittable"
                                    propertyName="storesKittableFlag"
                                    allowNoValue
                                    items={[
                                        { id: 'Y', displayText: 'Yes' },
                                        { id: 'N', displayText: 'No' }
                                    ]}
                                    onChange={handleFieldChange}
                                />
                            </Grid>
                            <Grid size={3}>
                                <InputField
                                    value={formValues.salesAccountId}
                                    type="number"
                                    fullWidth
                                    label="Stores Kit Priority"
                                    propertyName="storesKittingPriority"
                                    onChange={handleFieldChange}
                                />
                            </Grid>
                            <Grid size={4}>
                                <InputField
                                    value={formValues.auditFrequencyWeeks}
                                    type="number"
                                    fullWidth
                                    label="Audit Freq (wks)"
                                    propertyName="auditFrequencyWeeks"
                                    onChange={handleFieldChange}
                                />
                            </Grid>
                            <Grid size={3}>
                                <Dropdown
                                    value={formValues.accessibleFlag}
                                    fullWidth
                                    label="Accessible"
                                    propertyName="accessibleFlag"
                                    allowNoValue
                                    items={[
                                        { id: 'Y', displayText: 'Yes' },
                                        { id: 'N', displayText: 'No' }
                                    ]}
                                    onChange={handleFieldChange}
                                />
                            </Grid> 
                            {!creating && <>
                                <Grid size={2}>
                                    <InputField
                                        value={formValues.locationId} 
                                        disabled                                    
                                        fullWidth
                                        label="Location Id"
                                        propertyName="locationId"
                                    />
                                </Grid>
                                <Grid size={3}>
                                    <InputField
                                        value={formValues.dateInvalid ? moment(formValues.dateInvalid).format('DD MMM YYYY') : ""} 
                                        disabled                                    
                                        fullWidth
                                        label="Date Invalid"
                                        propertyName="locationId"
                                    />
                                </Grid>                            
                                <Grid size={7}>
                                    {!formValues.dateInvalid && <Button onClick={makeInvalid}>Make Invalid</Button>}
                                </Grid>                            
                            </>}  
                            <Grid size={12}>
                                <SaveBackCancelButtons
                                    backClick={() => navigate('/stores2/storage')}
                                    saveClick={() => {
                                        setChangesMade(false);

                                        if (creating) {
                                            createLocation(null, formValues);
                                        } else {
                                            updateLocation(id, formValues);
                                        }
                                    }}
                                    saveDisabled={!changesMade || !okToSave()}
                                    cancelClick={() => {
                                        setChangesMade(false);
                                        if (creating) {
                                            setFormValues({ countryCode: 'GB' });
                                        } else {
                                            setFormValues(locationGetResult);
                                        }
                                    }}
                                />
                            </Grid>
                            {!creating && (
                                <Grid size={12}>
                                    <Typography variant="body2">{lastAuditInfo()}</Typography>
                                </Grid>
                            )}
                        </>
                    )
                )}
            </Grid>
        </Page>
    );
}

StorageLocation.propTypes = { creating: PropTypes.bool };
StorageLocation.defaultProps = { creating: false };

export default StorageLocation;
